test/Test/Pos/Ssc/GodTossing/Toss/BaseSpec.hs

```diff
@@ -0,0 +1,573 @@
+        "unsuccessfully checks payload of certificates with the right exception for each\
+        \ failure case"
+
+verifiesOkComm :: CommitmentOpening -> Bool
+verifiesOkComm CommitmentOpening{..} =
+    verifyCommitment coCommitment
+
+notVerifiesBadComm :: BadCommitment -> Bool
+notVerifiesBadComm (getBadComm -> badComm) =
+    not . verifyCommitment $ badComm
+
+verifiesOkCommSig :: SecretKey -> Commitment -> EpochIndex -> Bool
+verifiesOkCommSig sk comm epoch =
+    let commSig = (toPublic sk, comm, sign SignCommitment sk (epoch, comm))
+    in verifyCommitmentSignature epoch commSig
+
+notVerifiesBadCommSig :: BadSignedCommitment -> EpochIndex -> Bool
+notVerifiesBadCommSig (getBadSignedC -> badSignedComm) epoch =
+    not $ verifyCommitmentSignature epoch badSignedComm
+
+verifiesOkOpening :: CommitmentOpening -> Bool
+verifiesOkOpening CommitmentOpening{..} =
+    verifyOpening coCommitment coOpening
+
+notVerifiesBadOpening :: BadCommAndOpening -> Bool
+notVerifiesBadOpening (getBadCAndO -> badCommsAndOp) =
+    not . uncurry verifyOpening $ badCommsAndOp
+
+emptyPayload
+    :: Monoid container
+    => (container -> ExceptT e PureToss a)
+    -> MultiRichmenStake
+    -> GtGlobalState
+    -> Bool
+emptyPayload pureToss mrs gtgs =
+    isRight $ tossRunner mrs gtgs $ pureToss mempty
+
+newtype GoodCommsPayload = GoodComms
+    { getGoodComms :: (EpochIndex, GtGlobalState, CommitmentsMap, MultiRichmenStake)
+    } deriving (Show, Eq)
+
+instance Arbitrary GoodCommsPayload where
+    arbitrary = GoodComms <$> do
+        -- These fields won't be needed for anything, so they can be entirely arbitrary.
+        _gsOpenings <- arbitrary
+        _gsShares <- arbitrary
+
+        -- The epoch used in the tests is generated separately to make sure it exists.
+        (epoch, m) <- arbitrary :: Gen (EpochIndex, MultiRichmenStake)
+        richmen <- getValid <$> (arbitrary :: Gen (ValidRichmenStake GenesisMpcThd))
+        let richmenIds = HM.keys richmen
+            mrs = HM.insert epoch richmen m
+
+        -- This is the list of richmen which will be used to make keys for commitments
+        -- in the argument 'CommitmentsMap'.
+        richmenWithCerts <- sublistOf richmenIds
+        richmenWithComms <- sublistOf richmenWithCerts
+
+        stableCerts <- HM.fromList <$>
+            mapM (\r -> (,) <$> pure r <*> (arbitrary :: Gen VssCertificate))
+                 richmenWithCerts
+
+        -- Only the simplest case is considered w.r.t. the 'lastKnownEoS' field:
+        -- 'checkCommitmentsPayload' is called with the same epoch as the 'lastKnownEoS'
+        -- in 'gsVssCertificate'.
+        -- This is because rolling back slots is and should be tested elsewhere.
+        _gsVssCertificates <- VssCertData
+            <$> (pure . EpochOrSlot . Right . crucialSlot $ epoch)
+            <*> pure stableCerts
+            <*> arbitrary
+            <*> arbitrary
+            <*> arbitrary
+            <*> arbitrary
```

Comment:
This reply is for this comment and the following 2 as well:
Since the purpose of these datatypes is just to test the `check*Payload` functions, does it matter if some unrelated invariants are violated? They're unrelated to the checks performed in those functions, and therefore to the functions themselves.

If in the future we change these checking functions so that these invariants become relevant, it will make sense to change these `Arbitrary` instances to reflect this. Otherwise, is it really necessary?